server/src/main/java/com/google/fhir/gateway/AuditEventHelperImpl.java

```diff
@@ -0,0 +1,362 @@
+        break;
+
+      case UPDATE:
+        auditEventList =
+            createAuditEventCore(
+                resources, BalpProfileEnum.PATIENT_UPDATE, BalpProfileEnum.PATIENT_UPDATE);
+        break;
+
+      case DELETE:
+
+        // NOTE: The success of processing of DELETE is heavily dependent on server validation
+        // policy e.g. If you have a permission list that references the resource being deleted,
+        // it breaks the delete operation. Likewise if you reference the resource in AuditEvent
+        // such as logging when you first created or updated it the operation breaks. In case of
+        // such an error, AuditEvent.outcome and AuditEvent.outcomeDesc are populated accordingly
+
+        // Also note, this implementation only fully captures that a specific resource was deleted
+        // and by whom but not the compartment owner.
+        // With no access to the actual deleted resource we can't get the compartment owner unless
+        // the resource itself is a Patient. A crude way to get the owner would be to fetch the
+        // actual resource from the database first before creating the AuditEvent.
+
+        // This implementation skips logging Deletes if the operation is conditional (thus no
+        // Resource ID present)
+
+        auditEventList =
+            createAuditEventCore(
+                resources, BalpProfileEnum.PATIENT_DELETE, BalpProfileEnum.BASIC_DELETE);
+        break;
+
+      default:
+        break;
+    }
+
+    // TODO Investigate bulk saving (batch processing) instead to improve performance. We'll
+    // probably need a mechanism for chunking e.g. 100, 200, or 500 batch
+    for (AuditEvent auditEvent : auditEventList) {
+      auditEvent.getPeriod().setEnd(new Date());
+      try {
+        this.httpFhirClient.postResource(auditEvent);
+      } catch (IOException exception) {
+        ExceptionUtil.throwRuntimeExceptionAndLog(logger, exception.getMessage(), exception);
+      }
+    }
+  }
+
+  private String getResourceTemplate(String resourceName, String resourceId) {
+    return String.format("{ \"resourceType\": \"%s\", \"id\": \"%s\"}", resourceName, resourceId);
+  }
+
+  private List<AuditEvent> createAuditEventCore(
+      List<IBaseResource> resources, BalpProfileEnum patientProfile, BalpProfileEnum basicProfile) {
+    List<AuditEvent> auditEventList = new ArrayList<>();
+
+    for (IBaseResource iBaseResource : resources) {
+      try {
+        Preconditions.checkState(iBaseResource instanceof DomainResource);
+
+        DomainResource resource = (DomainResource) iBaseResource;
+
+        Set<String> patientIds = patientFinder.findPatientIds(resource);
+
+        if (!patientIds.isEmpty()) {
+          auditEventList.add(createAuditEvent(resource, patientProfile, patientIds));
+        } else {
+          auditEventList.add(createAuditEvent(resource, basicProfile, Set.of()));
```

Comment:
nit: While it general I am a fan of small functions but in this case adding the extra call of `createAuditEvent` does not help readability (may even hinder it). It only does a branch based on `resource instanceof OperationOutcome`; I would suggest to do that directly here and get rid of the empty set argument altogether.

Reply:
Yeah I suppose we can go with that approach. Its two things we need to take care of actually:
1. Branch based on whether the patientIds is empty
2. Passing the correct profile to the `createAuditEvent` function

But I see you mentioned `resource instanceof OperationOutcome` as the condition. Did you intend for this comment to be for another line?

Reply:
I've refactored to move the logic to the calling site above